plotly/dashboard_objs/dashboard_objs.py

```diff
@@ -0,0 +1,391 @@
+
+    index_for_new_box = dashboard_html.find('</script>') - 1
+    dashboard_html = (dashboard_html[:index_for_new_box] + html_box +
+                      dashboard_html[index_for_new_box:])
+    return dashboard_html
+
+
+def add_html_text(dashboard_html, text, top_left_x, top_left_y, box_w, box_h):
+    """
+    Add a number to the middle of an HTML box.
+    """
+    html_text = """<!-- Insert box numbers -->
+          context.font = '{font_size}pt Times New Roman';
+          context.textAlign = 'center';
+          context.fillText({text}, {top_left_x} + 0.5*{box_w}, {top_left_y} + 0.5*{box_h});
+    """.format(text=text, top_left_x=top_left_x, top_left_y=top_left_y,
+               box_w=box_w, box_h=box_h, font_size=font_size)
+
+    index_to_add_text = dashboard_html.find('</script>') - 1
+    dashboard_html = (dashboard_html[:index_to_add_text] + html_text +
+                      dashboard_html[index_to_add_text:])
+    return dashboard_html
+
+
+class Dashboard(dict):
+    def __init__(self, content=None):
+        if content is None:
+            content = {}
+
+        self.box_ids_to_path = {}
+        if not content:
+            self['layout'] = _empty_box()
+            self['version'] = 2
+            self['settings'] = {}
+        else:
+            self['layout'] = content['layout']
+            self['version'] = content['version']
+            self['settings'] = content['settings']
+
+            self._assign_boxes_to_ids()
+
+    def _assign_boxes_to_ids(self):
+        self.box_ids_to_path = {}
+        all_nodes = []
+        node_gen = node_generator(self['layout'])
+
+        finished_iteration = False
+        while not finished_iteration:
+            try:
+                all_nodes.append(node_gen.next())
+            except StopIteration:
+                finished_iteration = True
+
+        for node in all_nodes:
+            if (node[1] != () and node[0]['type'] == 'box'
+                    and node[0]['boxType'] != 'empty'):
+                try:
+                    max_id = max(self.box_ids_to_path.keys())
+                except ValueError:
+                    max_id = 0
+                self.box_ids_to_path[max_id + 1] = list(node[1])
+
+    def _insert(self, box_or_container, path):
+        """Performs user-unfriendly box and container manipulations."""
+        if any(first_second not in ['first', 'second'] for first_second in path):
+            raise exceptions.PlotlyError(
+                "Invalid path. Your 'path' list must only contain "
+                "the strings 'first' and 'second'."
+            )
+
+        if 'first' in self['layout']:
+            loc_in_dashboard = self['layout']
+            for index, first_second in enumerate(path):
+                if index != len(path) - 1:
+                    loc_in_dashboard = loc_in_dashboard[first_second]
+                else:
+                    loc_in_dashboard[first_second] = box_or_container
+
+        else:
```

Comment:
Am I being unclear with `first_second`? That variable more accurately means `first_or_second`, so it can take values `first` or `second`.

On another or related note, `dashboard['layout']['second']` is *always* an emptyBox, so no path will ever look like `('second', 'second', 'first', ...)`